### PR TITLE
[WIP / for discussion] Lazy datasets

### PR DESCRIPTION
**Pull request by joelgrus**
here's a proposal for how to handle lazy datasets

at a high level, here are the changes

* make `Dataset.instances` private (and eliminate all external access to it)
* replace it with `Dataset.iterinstances() -> Iterator[Instance]` and change all calling code to use that
* introduce `Dataset.num_instances` property to replace calls to `len(dataset.instances)`
* create `LazyDataset` subclass that takes an "instance generator" instead of a list of instances
* create lazy `DataIterator` subclass that doesn't rely on having all instances in-memory
* modify `commands/train` to build the vocabulary from a `LazyDataset` instead of a `Dataset`
* add some tests

I'm not certain this is the right approach. It's a breaking change for anyone who's using `Dataset.instances` directly, although I'm not sure if that's anyone or not. A non-breaking alternative would have `LazyDataset` not be a subclass of `Dataset` and then try to flow those changes through, but that felt worse.

Some of the steps around how the LazyDataset handles indexing and padding are a little bit clunky, but I think they're right.

I don't *think* this should have any performance impacts (I think now during training we'll make a shallow copy of the list of instances each epoch, but that should be really fast).

Let me know what you think about this approach / the naming.

### CODE REVIEW

**Comment by joelgrus** [comment]
I also strongly considered changing `Vocabulary.from_dataset` to `Vocabulary.from_instances` (it would have saved the trouble of creating a fake lazy dataset in `commands.train`, which is expensive if the underlying datasets are lazy (since it has to iterate over all of them)), but I ran into some nasty circular import problems, and eventually I gave up.

But if that's a better approach, I can figure that problem out.

File: allennlp/data/lazy_dataset.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            instance_lengths = instance.get_padding_lengths()
+            for field_name, instance_field_lengths in instance_lengths.items():
+                for padding_key, length in instance_field_lengths.items():
+                    prev_max_length = padding_lengths[field_name].get(padding_key, 0)
+                    padding_lengths[field_name][padding_key] = max(length, prev_max_length)
+
+        self.padding_lengths = {**padding_lengths}
+
+    def iterinstances(self) -> Iterator[Instance]:
+        for i, instance in tqdm.tqdm(enumerate(self.generator())):
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/lazy_dataset.py]: Ah, actually, the way that you're using `LazyDataset` for vocab creation means it doesn't make sense to crash or give a warning here in that case.  Hmm...

File: tests/data/iterators/lazy_iterator_test.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+from allennlp.common import Params
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.data import LazyDataset, Instance, Token, Vocabulary
+from allennlp.data.fields import TextField
+from allennlp.data.iterators import LazyIterator
+from allennlp.data.token_indexers import SingleIdTokenIndexer
+
+class LazyIteratorTestCase(AllenNlpTestCase):
+    def setUp(self):
```
Comment (matt-gardner) [inline tests/data/iterators/lazy_iterator_test.py]: Why not just put these methods in `TestLazyIterator`?

File: allennlp/data/lazy_dataset.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+
+class LazyDataset(Dataset):
+    """
+    A collection of :class:`~allennlp.data.instance.Instance` objects.
+    The ``Instances`` have ``Fields``, and the fields
+    could be in an indexed or unindexed state - the ``Dataset`` has methods around indexing the
+    data and converting the data into arrays.
+    """
+    def __init__(self, generator: Callable[[], Iterator[Instance]]) -> None:
```
Comment (matt-gardner) [inline allennlp/data/lazy_dataset.py]: I think this generator needs a little bit of documentation - particularly, you should note that the generator needs to run forever, looping over the data when it needs to.

File: tests/data/iterators/lazy_iterator_test.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        # We just want to get the single-token array for the text field in the instance.
+        instances = [tuple(instance.data.cpu().numpy())
+                     for batch in batches
+                     for instance in batch['text']["tokens"]]
+        assert len(instances) == 5
+        self.assert_instances_are_correct(instances)
+
+    def test_call_iterates_over_data_forever(self):
+        generator = LazyIterator(batch_size=2)(self.dataset)
+        batches = [next(generator) for _ in range(18)]  # going over the data 6 times
```
Comment (matt-gardner) [inline tests/data/iterators/lazy_iterator_test.py]: I'm trying to figure out where in the code you get the infinite looping behavior, but I can't find it.  What am I missing?

File: allennlp/commands/train.py [hunk 5]
```diff
@@ -152,10 +152,18 @@ def train_model(params: Params, serialization_dir: str) -> Model:
...
+    # As the individual Datasets may be lazy (and, in particular, large), we don't want to
+    # manifest a List of their instances to construct the vocabulary. Instead we use a LazyDataset
+    # with the combined instances.
+    def combined_generator():
+        yield from (instance for key, dataset in all_datasets.items()
+                    for instance in dataset.iterinstances()
+                    if key in datasets_for_vocab_creation)
+
+    lazy_combined_dataset = LazyDataset(combined_generator)
+    vocab = Vocabulary.from_params(params.pop("vocabulary", {}), lazy_combined_dataset)
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/commands/train.py]: Or, if you really do want our code to do the vocab computation with a super large dataset, this still does the right thing, iterating over the data once to compute the vocab, and iterating over it once again during training.

File: allennlp/data/lazy_dataset.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/lazy_dataset.py]: Also, as I noted above, it's also reasonable to use our code for vocab creation with a large dataset, so, yeah, you don't want to crash here.  Ignore this.

File: allennlp/data/iterators/lazy_iterator.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    ----------
+    batch_size : int, optional, (default = 32)
+        The size of each batch of instances yielded when calling the iterator.
+    """
+    def __init__(self, batch_size: int = 32) -> None:
+        self._batch_size = batch_size
+
+    def _yield_one_epoch(self, dataset: Dataset, shuffle: bool, cuda_device: int, for_training: bool):
+        if shuffle:
+            logger.warning("shuffle is not implemented for LazyIterator")
```
Comment (nalourie-ai2) [inline allennlp/data/iterators/lazy_iterator.py]: This seems liable to create bugs in the future, since we might use this private method in some generic way that requires shuffle and not notice the message in the logs. If this is a private method perhaps it should throw an error instead?

Comment (matt-gardner) [reply allennlp/data/iterators/lazy_iterator.py]: Yeah, a `ConfigurationError` does seem warranted here, as this would be the result of mis-configuring something in your json file.

File: allennlp/data/dataset.py [hunk 7]
```diff
@@ -38,26 +38,33 @@ class that call the constructor, such as `truncate()`.
...
         """
         Converts all ``UnindexedFields`` in all ``Instances`` in this ``Dataset`` into
         ``IndexedFields``.  This modifies the current object, it does not return a new object.
         """
         logger.info("Indexing dataset")
-        for instance in tqdm.tqdm(self.instances):
+        for instance in tqdm.tqdm(self._instances):
             instance.index_fields(vocab)
 
+    def iterinstances(self) -> Iterator[Instance]:
```
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/data/dataset.py]: that's a good idea

File: allennlp/data/iterators/lazy_iterator.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            yield batch.as_tensor_dict(padding_lengths,
+                                       cuda_device=cuda_device,
+                                       for_training=for_training)
+
+    def _create_batches(self, dataset: Dataset, shuffle: bool) -> List[List[Instance]]:
+        raise RuntimeError("should never be called")
+
+    @overrides
+    def get_num_batches(self, dataset: Dataset) -> int:
+        return math.ceil(dataset.num_instances / self._batch_size)
```
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/data/iterators/lazy_iterator.py]: but `num_instances` has to come from somewhere? there's no way to get it without iterating over the dataset, so where does that iteration happen?

File: allennlp/data/lazy_dataset.py (hunk 3 above)
Comment (joelgrus) [reply allennlp/data/lazy_dataset.py]: well, that's not how it works, so it definitely needs more documentation! 😀

each call to `generator()` needs to loop through the data once, but `generator` has to be able to be called multiple times.

File: tests/data/iterators/lazy_iterator_test.py (hunk 4 above)
Comment (joelgrus) [reply tests/data/iterators/lazy_iterator_test.py]: in the base class, `__call__` just does

```
while True:
  yield from self._yield_one_epoch(...)
```

and each call to `LazyIterator._yield_one_epoch()` calls `self.generator()` to get a fresh iterator over the whole dataset.

File: tests/data/iterators/lazy_iterator_test.py (hunk 2 above)
Comment (joelgrus) [reply tests/data/iterators/lazy_iterator_test.py]: because I copied the `BasicIterator` tests and that's how they were written 😄

File: allennlp/data/iterators/lazy_iterator.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/iterators/lazy_iterator.py]: It's provided by the caller, in the configuration file.  Like, if you have a huge language modeling dataset, you know how many instances you want to use for each "epoch", and it doesn't really matter whether you're actually iterating over the whole dataset or not.  `num_instances` here no longer necessarily has any relation to the actual size of the data, it's just a specification for how frequently you want to check in on progress by running validation.  If you know the actual number of instances in the data, and want to specify that in the configuration file, that's fine too, but for huge datasets where lazy creation is needed, you likely want to do validation more frequently than waiting for a full pass over the whole data.

File: tests/data/iterators/lazy_iterator_test.py (hunk 2 above)
Comment (matt-gardner) [reply tests/data/iterators/lazy_iterator_test.py]: Ah, right, those actually do get used in other places, because the other iterators are so similar.

File: tests/data/iterators/lazy_iterator_test.py (hunk 4 above)
Comment (matt-gardner) [reply tests/data/iterators/lazy_iterator_test.py]: Ah, great, thanks.  So, we need to be able to handle cases where one "epoch" does not exhaust the generator, and the next epoch just keeps going from where the last epoch left off.  This is pretty important, and related to my [most recent comment above](https://github.com/allenai/allennlp/pull/660#discussion_r159767396).  The iterator should have a `num_instances_per_epoch` parameter, or something (maybe `num_batches_per_epoch` instead...), and that's used here.  And then when the generator is exhausted, you start it again.  I think there are a few ways to get this kind of behavior, and I'm not sure what the best one is.

File: allennlp/data/lazy_dataset.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        """
+        A LazyDataset just takes a way of generating instances.
+        """
+        # Call superclass constructor with no instances, we'll override the methods that use them.
+        super().__init__([])
```
Comment (DeNeutoy) [inline allennlp/data/lazy_dataset.py]: I don't like this much - see my comment below about possibly removing this whole class and supporting laziness in `Dataset` by passing a flag.

**Comment by matt-gardner** [comment]
I also wasn't too thrilled with the `NotImplementedError` methods.  I missed the call to the superclass constructor, but it's the same problem.  I was considering suggesting making `Dataset` an abstract API, with concrete `InMemoryDataset` and `LazyDataset` classes, where `InMemoryDataset` defines the methods that `LazyDataset` doesn't need.  It seemed a little overboard to me, though.

The trouble with passing in the vocab to the `Dataset` is that the `DatasetReader` (which builds the `Dataset`) can't have access to it - it hasn't been built yet.  You could maybe get around this by having the `DataIterator` handle indexing, as you suggest.

I think the fundamental difference between lazy and non-lazy datasets, in terms of the computation done, is that non-lazy datasets can index all of the data once and store the result in memory, saving computation when iterating multiple times over the data.  In fact, that's probably the _reason_ that we need to be lazy in the first place - the text representing the data probably doesn't take all that much memory, but the indexed tensors do.  Everything else is largely the same - lazy datasets will (probably) read from disk while they're iterating over the data, so it'll be slower, but the basic algorithm for going over the data is the same.  

So, this is a long way of saying that the difference in behavior in `index_instances` (and the fact that `iterinstances()` does indexing in the `LazyDataset`) is kind of the _definition_ of the difference between lazy and non-lazy datasets.  I think the right thing to do here is to update documentation for what these methods do, instead of trying to make them behave the same.  `index_instances` could be defined as "Makes it so that instances returned when iterating over this dataset have already been indexed using the given vocabulary.  _When_ the indexing computation happens is determined by the concrete `Dataset` class."

**Comment by joelgrus** [comment]
I think it will be ugly to have a `lazy=True` flag, the internals are very different and you'll just have lots of if/then blocks everywhere. Maybe an abstract base class makes more sense

**Comment by DeNeutoy** [comment]
If we're going to move to an abstract base class it might be worth investigating how easily we can use pytorch's `Dataset`, I don't know if that even gets us anything - I think they have some good multi-threading stuff we might want to use eventually, but that might be `DataLoaders` instead.

**Comment by matt-peters** [comment]
Implementation wise, here's my experience in building an efficient lazy dataset loader for another project.  This is for datasets larger then machine memory.  I don't know how easy/hard this would be to implement in the proposed APIs (as I'm having a hard time following the discussion without detailed knowledge of the current API).

On disk, the dataset is split into `N` random shards, where `N` might be 100-10,000.  Each shard can comfortably fit into memory.  To iterate through the data, the data loader picks a random shard, then treats it identically to a small dataset.  It randomly shuffles the shard, batches according to sequence length, indexes and places the resulting tensors onto a Queue.  This all happens in a separate process.  Then main modeling process then simply pulls batches from the Queue.  I found that using pytorch's monkey patched `multiprocessing` was much faster then the native Python one for tensors (e.g. `from torch.multiprocessing import Queue, Process`).  In this way, there is almost no overhead in the master process to read the data.

**Comment by matt-gardner** [comment]
Yeah, @matt-peters, those are some good points.  I had in my head that in the future we would probably want the `LazyIterator` to actually load some larger number of instances, then use something like our `BucketIterator` to go over the loaded instances before moving on.  I didn't say anything because I _think_ it's orthogonal to how the `LazyDataset` is designed, but it's worth keeping in mind - maybe there's a simpler way to put all of this together than what we're thinking of.

**Comment by joelgrus** [comment]
I uploaded some changes, but I'm still a little bit stuck, I'll talk about it at the meeting tomorrow
